ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl/Types.hs

```diff
@@ -268,36 +269,46 @@ validateIS
   -> MempoolCapacityBytesOverride
   -> ValidationResult (Validated (GenTx blk)) blk
 validateIS istate lstate lconfig capacityOverride =
-    validateStateFor capacityOverride lconfig (ForgeInUnknownSlot lstate) istate
+    fst $ validateStateFor capacityOverride lconfig (ForgeInUnknownSlot lstate) istate
 
 -- | Given a (valid) internal state, validate it against the given ledger
 -- state and 'BlockSlot'.
 --
 -- When these match the internal state's 'isTip' and 'isSlotNo', this is very
 -- cheap, as the given internal state will already be valid against the given
--- inputs.
+-- inputs. The state passed as parameter potentially includes some values which
+-- were not required by the set of transactions that were used on the last
+-- mempool revalidation/sync. Therefore the way to get an state with those new
+-- values is to apply the same accumulated differences to the new state.
 --
--- When these don't match, the transaction in the internal state will be
--- revalidated ('revalidateTxsFor').
+-- When these don't match, the transactions in the internal state will be
+-- revalidated ('revalidateTxsFor') on top of the given ledger state.
+--
+-- The returned ticked ledger state will be used by the block forging logic to
+-- determine how many transactions can be put on the new block, hence, the 'mk'
+-- is irrelevant.
 validateStateFor
   :: (LedgerSupportsMempool blk, HasTxId (GenTx blk), ValidateEnvelope blk)
   => MempoolCapacityBytesOverride
   -> LedgerConfig     blk
-  -> ForgeLedgerState blk ValuesMK
+  -> ForgeLedgerState blk
   -> InternalState    blk
-  -> ValidationResult (Validated (GenTx blk)) blk
+  -> (ValidationResult (Validated (GenTx blk)) blk, TickedLedgerState blk TrackingMK)
 validateStateFor capacityOverride cfg blockLedgerState is
     | isTip    == castHash (getTipHash st')
     , isSlotNo == slot
-    = validationResultFromIS is
+    = (validationResultFromIS
+        is { isLedgerState = reapplyTrackingTicked (isLedgerState is) (forgeLedgerState blockLedgerState) }
+      , isLedgerState is)
```

Comment:
Add a comment mentioning that `isLedgerState is == st' (+) txs (-) newValues`, but as the `mk` is irrelevant, for our purpose it is the same and we avoid computing `st'`.

Reply:
Done